src/Ouroboros/Network/ChainFragment.hs

```diff
@@ -0,0 +1,561 @@
+-- * The 'blockSlot' of @bSucc@ must be strictly larger than that of @b@.
+-- * The 'blockNo' of @bSucc@ must be 1 greater than that of @b@.
+--
+-- This function does not check whether any of the two blocks satisfy
+-- 'blockInvariant'.
+isValidSuccessorOf :: HasHeader block
+                   => block  -- ^ @bSucc@
+                   -> block  -- ^ @b@
+                   -> Bool
+isValidSuccessorOf bSucc b =
+    pointHash p == blockPrevHash bSucc
+ && pointSlot p <  blockSlot bSucc
+ && blockNo   b == pred (blockNo bSucc)
+  where
+    p = blockPoint b
+
+validExtension ::  HasHeader block => ChainFragment block -> block -> Bool
+validExtension c bSucc =
+    blockInvariant bSucc
+ && blockSlot bSucc /= Slot 0
+ && case head c of
+      Nothing -> True
+      Just b  -> bSucc `isValidSuccessorOf` b
+
+head :: HasHeader block => ChainFragment block -> Maybe block
+head (_ :> b) = Just b
+head Empty    = Nothing
+
+headPoint :: HasHeader block => ChainFragment block -> Maybe (Point block)
+headPoint = fmap blockPoint . head
+
+headSlot :: HasHeader block => ChainFragment block -> Maybe Slot
+headSlot = fmap pointSlot . headPoint
+
+headHash :: HasHeader block => ChainFragment block -> Maybe (Hash block)
+headHash = fmap pointHash . headPoint
+
+headBlockNo :: HasHeader block => ChainFragment block -> Maybe BlockNo
+headBlockNo = fmap blockNo . head
+
+
+-- | Make a list of blocks from a 'ChainFragment', in newest-to-oldest order.
+toNewestFirst :: HasHeader block => ChainFragment block -> [block]
+toNewestFirst = foldChainFragment (flip (:)) []
+
+-- | Make a list of blocks from a 'ChainFragment', in oldest-to-newest order.
+toOldestFirst :: HasHeader block => ChainFragment block -> [block]
+toOldestFirst (ChainFragment ft) = Foldable.toList ft
+
+-- | Make a 'ChainFragment' from a list of blocks in newest-to-oldest order.
+fromNewestFirst :: HasHeader block => [block] -> ChainFragment block
+fromNewestFirst bs = foldr (flip (:>)) Empty bs
+
+-- | Make a 'ChainFragment' from a list of blocks in oldest-to-newest order.
+fromOldestFirst :: HasHeader block => [block] -> ChainFragment block
+fromOldestFirst bs = assert (valid c) c
+  where
+    c = ChainFragment $ FT.fromList bs
+
+-- | Drop the newest @n@ blocks from the 'ChainFragment'.
+dropNewest :: HasHeader block
+           => Int  -- ^ @n@
+           -> ChainFragment block -> ChainFragment block
+dropNewest n cf@(ChainFragment c) =
+    ChainFragment $ FT.takeUntil (\v -> bmSize v > remainingLength) c
+  where
+    remainingLength = length cf - n
+
+-- | Drop the oldest @n@ blocks from the 'ChainFragment'.
+dropOldest :: HasHeader block
+           => Int  -- ^ @n@
+           -> ChainFragment block -> ChainFragment block
+dropOldest n (ChainFragment c) =
+    ChainFragment $ FT.dropUntil (\v -> bmSize v > n) c
+
+
+length :: ChainFragment block -> Int
+length (ChainFragment c) = Foldable.length c
+
+null :: ChainFragment block -> Bool
+null (ChainFragment c) = Foldable.null c
```

Comment:
Good catch. I have re-implemented `length` using the measure to obtain a O(1) complexity. I have also re-implemented `null` in terms of `FT.null`, which is documented to have O(1), instead of relying on the undocumented complexity of `Foldable.null`.